FWCore/Framework/interface/GenericProduct.h

```diff
@@ -0,0 +1,104 @@
+#ifndef FWCore_Framework_interface_GenericProduct_h
+#define FWCore_Framework_interface_GenericProduct_h
+
+#include <cstring>
+#include <memory>
+#include <typeinfo>
+#include <utility>
+
+#include "DataFormats/Common/interface/OrphanHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Reflection/interface/ObjectWithDict.h"
+#include "FWCore/Reflection/interface/TypeWithDict.h"
+
+namespace edm {
+
+  class GenericProduct {
+  public:
+    GenericProduct(ObjectWithDict const& object, TypeWithDict const& wrapper)
+        : object_{object}, wrapperType_{wrapper} {}
+
+    // The GenericProduct *does not* take ownership of the object.
+    // The caller must guarantee that the object remains valid as long as needed.
+    template <typename T>
+    GenericProduct(T& object) : object_{typeid(T), &object}, wrapperType_{typeid(Wrapper<T>)} {}
+
+    ObjectWithDict const& object() const { return object_; }
+    TypeWithDict const& wrapperType() const { return wrapperType_; }
+
+  private:
+    ObjectWithDict object_;
+    TypeWithDict wrapperType_;
```

Comment:
Is there an efficient way to build the `TypeWithDict` for `Wrapper<T>` from a `TypeWithDict` for `T` ?

If there was, we could use `ObjectWithDict` directly, without the need to introduce the `GenericProduct` type.